### PR TITLE
Fix RemoveRelativePathComponents (again) and fix mod security with paths starting with ".."

### LINKED ISSUE #4909: Mod security: Still broken with different --world
`./minetest --server --world ../testdir --config /dev/null`
<pre>
2016-12-14 19:48:52: ERROR[Main]: ModError: Failed to load and run script from minetest/bin/../builtin/init.lua:
2016-12-14 19:48:52: ERROR[Main]: Attempt to access external file ../testdir/auth.txt with mod security on.
2016-12-14 19:48:52: ERROR[Main]: stack traceback:
2016-12-14 19:48:52: ERROR[Main]: 	[C]: in function 'open'
2016-12-14 19:48:52: ERROR[Main]: 	minetest/bin/../builtin/game/auth.lua:37: in function 'read_auth_file'
2016-12-14 19:48:52: ERROR[Main]: 	minetest/bin/../builtin/game/auth.lua:82: in main chunk
2016-12-14 19:48:52: ERROR[Main]: 	[C]: in function 'dofile'
2016-12-14 19:48:52: ERROR[Main]: 	minetest/bin/../builtin/game/init.lua:24: in main chunk
2016-12-14 19:48:52: ERROR[Main]: 	[C]: in function 'dofile'
2016-12-14 19:48:52: ERROR[Main]: 	minetest/bin/../builtin/init.lua:38: in main chunk
</pre>

### PR DESCRIPTION
**Pull request by ShadowNinja**
Fixes #4909.
  * Fix RRPC: This used to return "/foo" for "../foo" when it should return the empty
string (i.e., error removing all relative components).  Also, "./foo" should be resolved to "foo".
  * Security: Trying to resolve a path with RRPC that can't be resolved without leaving leading parent components (e.g. "../worlds/foo" or "bar/../../worlds/foo") will fail.  To work around this, we leave the relative components and simply remove the trailing components one at a time, and bail out when we find a parent component.  This will still fail for paths like "worlds/foo/noexist/../auth.txt" (the path before the last parent component must not exist), but this is fine since  you won't be able to open a file with a path like that anyways (the O.S. will determine that the path doesn't exist.  Try `cat /a/../etc/passwd`).

### CODE REVIEW

File: src/filesys.cpp [hunk 1]
```diff
@@ -617,48 +617,52 @@ std::string RemoveRelativePathComponents(std::string path)
...
+			if (component_start == 0) {
+				// We need to remove the delemiter too
+				path = path.substr(component_with_delim_end, std::string::npos);
+			} else {
+				path = path.substr(0, pos) + DIR_DELIM +
+					path.substr(component_with_delim_end, std::string::npos);
+			}
+			if (pos > 0) {
+				pos++;
+			}
```
Comment (sfan5) [inline src/filesys.cpp]: brace removal possible

**Comment by paramat** [comment]
Does this replace #4849 ? If there's a conflict which should be merged first?

**Comment by sfan5** [comment]
@paramat No, those two PRs fix two seperate issues with mod security.

File: src/filesys.cpp (hunk 1 above)
Comment (est31) [inline src/filesys.cpp]: The code style nowhere mandates that braces should be removed. Both placing braces and not placing them is fine.

**Comment by ShadowNinja** [comment]
@sfan5: Done.

**Comment by paramat** [comment]
Applying: Fix RemoveRelatvePathComponents
Applying: Security: Fix resolving of some relative paths
error: patch failed: src/script/cpp_api/s_security.cpp:340
error: src/script/cpp_api/s_security.cpp: patch does not apply
Patch failed at 0002 Security: Fix resolving of some relative paths

When attempting to merge after #4849, so i will merge #4849 only for now.

**Comment by paramat** [comment]
#4849 merged.

**Comment by Zeno-** [comment]
merged as 

0f0502109eac44128e87906fff30b5d049392f1d
and
f522e7351a1eaffcd4b0f1f06fab65a44281f972

Both commits seemed large(ish) so I did not squash

**Pull request closed without merging**